jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ExtUtils.java

```diff
@@ -35,10 +65,186 @@ public static String toUri(JarEntryFile jarEntryFile) {
+        String handleId = uri.getQuery();
+        IPackageFragmentRoot packageRoot = (IPackageFragmentRoot) JavaCore.create(handleId);
+        String path = uri.getPath();
+
+        // if the file exists in the java packages
+        String[] segments = StringUtils.split(path, "/");
+        String packageName = StringUtils.join(Arrays.asList(segments).subList(0, segments.length - 1), '.');
+        IPackageFragment packageFragment = packageRoot.getPackageFragment(packageName);
+        if (packageFragment != null && packageFragment.exists()) {
+            Object[] objs = packageFragment.getNonJavaResources();
+            for (Object obj : objs) {
+                if (obj instanceof IJarEntryResource) {
+                    IJarEntryResource child = (IJarEntryResource) obj;
+                    if (child instanceof JarEntryFile && child.getFullPath().toPortableString().equals(path)) {
+                        return child;
+                    }
+                }
+
+            }
+        }
+        Object[] resources = packageRoot.getNonJavaResources();
+
+        for (Object resource : resources) {
+            if (resource instanceof JarEntryFile) {
+                JarEntryFile file = (JarEntryFile) resource;
+                if (file.getFullPath().toPortableString().equals(path)) {
+                    return file;
+                }
+            }
+            if (resource instanceof JarEntryDirectory) {
+                JarEntryDirectory directory = (JarEntryDirectory) resource;
+                return findFileInJar(directory, path);
+            }
+        }
+        return null;
+    }
+
+    public static JarEntryFile findFileInJar(JarEntryDirectory directory, String path) {
+        for (IJarEntryResource child : directory.getChildren()) {
+            if (child instanceof JarEntryFile && child.getFullPath().toPortableString().equals(path)) {
+                return (JarEntryFile) child;
+            }
+            if (child instanceof JarEntryDirectory) {
+                JarEntryFile file = findFileInJar((JarEntryDirectory) child, path);
+                if (file != null) {
+                    return file;
+                }
+            }
+        }
+        return null;
+    }
+
     public static IPath removeProjectSegment(String projectElementName, IPath path) {
         if (projectElementName.equals(path.segment(0))) {
             return path.removeFirstSegments(1).makeRelative();
         }
         return path;
     }
+
+    public static PackageNode createNodeForProject(IJavaElement javaElement) {
+        IProject proj = javaElement.getJavaProject().getProject();
+        PackageNode projectNode = new PackageNode(proj.getName(), proj.getFullPath().toPortableString(), NodeKind.PROJECT);
+        projectNode.setUri(proj.getLocationURI().toString());
+        return projectNode;
+    }
+
+    public static PackageNode createNodeForResource(IResource resource) {
```

Comment:
Please keep the resource operation in well organized manner.

Reply:
See the item 1 of Effecitve Java  3rd for your reference:  https://www.safaribooksonline.com/library/view/effective-java-3rd/9780134686097/ch2.xhtml#lev1